module/zfs/dmu_object.c

```diff
@@ -113,9 +186,8 @@ dmu_object_alloc_dnsize(objset_t *os, dmu_object_type_t ot, int blocksize,
 		 * dmu_tx_assign(), but there is currently no mechanism
 		 * to do so.
 		 */
-		(void) dnode_hold_impl(os, object, DNODE_MUST_BE_FREE, dn_slots,
-		    FTAG, &dn);
-		if (dn)
+		(void) dnode_alloc_impl(os, &object, dn_slots, FTAG, &db);
```

Comment:
if we're always going to ignore the return value, make it return `null`.  But see below comment about recommended return values.